### ISSUE TITLE
Support extensions in stack.yaml

### ISSUE THREAD

**Issue by ndmitchell** [commercialhaskell/stack#3137]
If you have multiple projects in a single stack.yaml, and want them to share the same list of default extensions, that's quite hard to do. You can specify:

```
ghc-options:
  "*": -XFoo
```

But that has two problems:

* It gets applied to packages you install as dependencies that are not in your project.
* It doesn't get applied when running `stack haddock`, so the extensions aren't reflected.

One solution would be adding an explicit `extensions` key which is plumbed through everywhere.

**Comment by ndmitchell** [comment]
The "best" workaround I've found is to use hpack specifications with `!include` in the package.yaml files, but it feels messy.

**Comment by mgsloan** [comment]
Hmm, I'm surprised that they wouldn't get applied when running `stack haddock`.  They ought to be, since it's run via cabal using the same configuration

This is certainly something worth considering, though I'm leaning towards an alternative solution - instead, have ghc-options allow specifying multiple packages.  So you could do

```
ghc-options:
  "myPackage1, myPackage2": -XFoo
```

**Comment by ndmitchell** [comment]
I really want `ghc-options` to have the ability to specify all local targets or all dependencies separately, e.g.

```
ghc-options:
   "$local": -XFoo
```

**Comment by mgsloan** [comment]
Seems like a reasonable feature, would likely merge a PR that added something like that.  Should consider the syntax carefully, though.

**Comment by snoyberg** [comment]
Is this resolved with the new `$locals` feature on master?

**Comment by mgsloan** [comment]
I think so!

**Comment by lexi-lambda** [comment]
I don’t think this is fixed resolved by the `$locals` change because options in the `ghc-options` block still don’t apply to `haddock`. That seems like maybe a separate issue, but it’s one I just ran into (adding `ghc-options: { some-package: '-XFoo' }` causes `some-package` to build but its Haddocks to fail).

**Comment by ndmitchell** [comment]
Agreed with @lexi-lambda - the work cited gets us closer, but not all the way there. Maybe there's a way to do the haddocks, but I suggest it's documented before closing this ticket, as it certainly isn't obvious.

**Comment by mgsloan** [comment]
I am not sure why haddock build would be any different, it is done via cabal after configure. So, this may be a bug with Cabal.  Can you please post a repository that acts as a straightforward reproduction of the problem?

**Comment by ndmitchell** [comment]
See https://github.com/ndmitchell/stack-3137. Two commits, first is the output of `stack new bob`, second requires MagicHash in the library and enables it in stack.yaml. Before both `stack build` and `stack haddock` work. After, `stack build` works, but `stack haddock` gives:

```
Running Haddock on library for bob-0.1.0.0..

src\Lib.hs:8:1: error:
    parse error (possibly incorrect indentation or mismatched brackets)
Haddock coverage:
  |
8 | someFunc# = putStrLn "someFunc"
  | ^

--  While building custom Setup.hs for package bob-0.1.0.0 using:
      C:\sr\setup-exe-cache\x86_64-windows\Cabal-simple_Z6RU0evB_2.0.1.0_ghc-8.2.2.exe --builddir=.stack-work\dist\5c8418a7 haddock --html --hoogle --html-location=../$pkg-$version/ --haddock-option=--hyperlinked-source
    Process exited with code: ExitFailure 1
```

**Comment by mgsloan** [comment]
@ndmitchell This appears to be a bug in Cabal that has been around for a long time (since 1.18) https://github.com/haskell/cabal/issues/4982 .  Or I just don't know how to pass ghc-options into haddock, but this seems like a bug.

So, options are to either fix Cabal or possibly stop relying on the Cabal library implementation of `haddock`.  I believe there were some other reasons to have our own haddock running code, but can't seem to find them with a quick search of the tracker.

**Comment by dbaynard** [comment]
@mgsloan I've just run into haddock issues — was the relevant haddock issue #729?
